src/Umbraco.Infrastructure/Persistence/Repositories/Implement/ContentRepositoryBase.cs

```diff
@@ -1130,12 +1136,24 @@ protected void PersistRelations(TEntity entity)
                 }
                 else
                 {
-                    relations.Add(new ReadOnlyRelation(entity.Id, id, relationTypeId));
+                    relations.Add((id, relationTypeId));
                 }
             }
 
-            // Save bulk relations
-            RelationRepository.SaveBulk(relations);
+            // Get all existing relations (optimize for adding new and keeping existing relations)
+            var query = Query<IRelation>().Where(x => x.ParentId == entity.Id).WhereIn(x => x.RelationTypeId, relationTypeLookup.Values);
+            var existingRelations = RelationRepository.GetPagedRelationsByQuery(query, 0, int.MaxValue, out _, null)
+                .ToDictionary(x => (x.ChildId, x.RelationTypeId)); // Relations are unique by parent ID, child ID and relation type ID
```

Comment:
I don't expect we'll have more than 2100 different relation types and run into a SQL parameter overflow here, so I've kept the query as simple as possible.